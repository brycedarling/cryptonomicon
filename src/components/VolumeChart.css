.volume-chart {
  font: 12px sans-serif;
}
