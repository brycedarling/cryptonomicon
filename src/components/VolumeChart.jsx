import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { scaleLinear, scaleTime } from 'd3-scale';
import { extent } from 'd3-array';
import { actions } from '../reducers';
import Margin from './Margin';
import Axis from './Axis';
import Line from './Line';

class VolumeChart extends Component {
  static propTypes = {
    width: PropTypes.number.isRequired,

    height: PropTypes.number.isRequired,

    margin: PropTypes.shape({
      bottom: PropTypes.number.isRequired,
      left: PropTypes.number.isRequired,
      right: PropTypes.number.isRequired,
      top: PropTypes.number.isRequired,
    }).isRequired,

    loadTicks: PropTypes.func.isRequired,

    data: PropTypes.arrayOf(PropTypes.shape({
      date: PropTypes.instanceOf(Date).isRequired,
      open: PropTypes.number.isRequired,
      high: PropTypes.number.isRequired,
      low: PropTypes.number.isRequired,
      close: PropTypes.number.isRequired,
      volume: PropTypes.number.isRequired,
      quoteVolume: PropTypes.number.isRequired,
      weightedAverage: PropTypes.number.isRequired,
    })).isRequired,
  };

  componentWillMount() {
    this.props.loadTicks();
  }

  get outerWidth() {
    const { margin } = this.props;
    return this.props.width + margin.left + margin.right;
  }

  get outerHeight() {
    const { margin } = this.props;
    return this.props.height + margin.top + margin.bottom;
  }

  get innerWidth() {
    const { margin } = this.props;
    return this.props.width - margin.left - margin.right;
  }

  get innerHeight() {
    const { margin } = this.props;
    return this.props.height - margin.top - margin.bottom;
  }

  get xScale() {
    return scaleTime()
      .domain(extent(this.props.data, d => d.date))
      .range([0, this.innerWidth]);
  }

  get yScale() {
    return scaleLinear()
      .domain(extent(this.props.data, d => d.volume / 1000000))
      .range([this.innerHeight, 0]);
  }

  render() {
    const { margin } = this.props;
    const {
      outerWidth, outerHeight, innerWidth, innerHeight, xScale, yScale,
    } = this;

    return (
      <svg
        className="volume-chart"
        width={outerWidth}
        height={outerHeight}
      >
        <Margin
          top={margin.top}
          left={margin.left}
          bottom={margin.bottom}
          right={margin.right}
        >
          <Axis
            label={{ text: 'Date', transform: `translate(${innerWidth / 2}, ${margin.bottom + 5})` }}
            className="x-axis"
            orientation="bottom"
            scale={xScale}
            transform={`translate(0, ${innerHeight})`}
          />
          <Axis
            label={{ text: 'Volume (in millions)', transform: `rotate(-90) translate(${-innerHeight / 2},${-margin.left + 15})` }}
            className="y-axis"
            orientation="left"
            scale={yScale}
          />
          <Line
            x={d => xScale(d.date)}
            y={d => yScale(d.volume / 1000000)}
            data={this.props.data}
          />
        </Margin>
      </svg>
    );
  }
}

const mapStateToProps = state => ({
  data: state.ticks.data,
});

function loadTicks() {
  return (dispatch, getState) => {
    const volumeChartUIState = getState().volumeChart.ui;

    const options = {
      currencyPair: volumeChartUIState.currencyPair,
      period: volumeChartUIState.period,
      start: volumeChartUIState.start,
    };

    if (volumeChartUIState.end) {
      options.end = volumeChartUIState.end;
    }

    actions.loadTicks(options)(dispatch);
  };
}

const mapDispatchToProps = {
  loadTicks,
};

export default connect(mapStateToProps, mapDispatchToProps)(VolumeChart);
