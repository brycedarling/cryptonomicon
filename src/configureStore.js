import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';

const configureStore = () => {
  const initialState = {};

  const enhancers = [];

  const middleware = [
    thunk,
  ];

  if (process.env.NODE_ENV === 'development') {
    const { __REDUX_DEVTOOLS_EXTENSION__ } = window;

    if (typeof devToolsExtension === 'function') {
      enhancers.push(__REDUX_DEVTOOLS_EXTENSION__());
    }
  }

  const composedEnhancers = compose(
    applyMiddleware(...middleware),
    ...enhancers,
  );

  const store = createStore(
    rootReducer,
    initialState,
    composedEnhancers,
  );

  return store;
};

export default configureStore;
