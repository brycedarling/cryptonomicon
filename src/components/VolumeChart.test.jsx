import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import VolumeChart from './VolumeChart';
import configureStore from '../configureStore';

describe('components', () => {
  describe('VolumeChart', () => {
    it('renders correctly', () => {
      const store = configureStore();

      const tree = renderer.create((
        <Provider store={store}>
          <VolumeChart
            width={600}
            height={270}
            margin={{
              top: 30, right: 20, bottom: 30, left: 50,
            }}
          />
        </Provider>
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
