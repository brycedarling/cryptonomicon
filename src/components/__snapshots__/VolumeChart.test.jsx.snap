// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components VolumeChart renders correctly 1`] = `
<svg
  className="volume-chart"
  height={330}
  width={670}
>
  <g
    className="margin"
    transform="translate(50, 30)"
  >
    <g
      className="axis x-axis"
      transform="translate(0, 210)"
    />
    <g
      className="axis y-axis"
      transform=""
    />
    <path
      className="line"
      d={null}
      style={
        Object {
          "fill": "none",
          "shapeRendering": "crispEdges",
          "stroke": "steelblue",
          "strokeWidth": 1,
        }
      }
    />
  </g>
</svg>
`;
